### PR TITLE
Todos - #1 View

### PR DESCRIPTION
**Pull request by matej494**

### CODE REVIEW

File: Marsiple/App/Todos/TodosViewController.swift [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        todosView.tableView.dataSource = self
+        todosView.tableView.delegate = self
+        view.addSubview(todosView)
+        todosView.snp.makeConstraints {
+            $0.edges.equalTo(view.safeAreaLayoutGuide)
+        }
+    }
+    
+    func setupTodos() -> [[Todo]] {
+        let uncompleated = [Todo](repeating: Todo(id: 1, title: "Todo 1", completed: false, userId: 1), count: 10)
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Typo `compleated` -> `completed`.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        let compleated = [Todo](repeating: Todo(id: 1, title: "Todo 2", completed: true, userId: 1), count: 10)
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Typo `compleated` -> `completed`.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        cell.updateProperties(withTodo: todos[indexPath.section][indexPath.row])
+        return cell
+    }
+}
+
+extension TodosViewController: UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        var removedTodo = todos[indexPath.section].remove(at: indexPath.row)
+        removedTodo.completed = !removedTodo.completed
+        todos[1 - indexPath.section].insert(removedTodo, at: todos[1 - indexPath.section].count)
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: That's an interesting approach (index calculation) 👍

**Review by dkulundzic — CHANGES_REQUESTED** [review]
Looks great 👍 Minor nitpicks and one question.

Do you plan on making the Todo completion animated in a later PR? The Todos should animate as they go through state changes.

**Comment by matej494** [comment]
I didn't know how to implement animation for this case so I planned to implement it later, but I've figured it out and it's coming in the next commit.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 4]
```diff
@@ -42,15 +42,18 @@ extension TodosViewController: UITableViewDataSource {
 
 extension TodosViewController: UITableViewDelegate {
     func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
-        var removedTodo = todos[indexPath.section].remove(at: indexPath.row)
-        removedTodo.completed = !removedTodo.completed
-        todos[1 - indexPath.section].insert(removedTodo, at: todos[1 - indexPath.section].count)
-        tableView.reloadData()
+        let finishIndexPath = IndexPath(row: todos[1 - indexPath.section].count, section: 1 - indexPath.section)
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: When `completing` a `Todo`, lets move its cell to the start of the `completed` section, not the end. The UX's much better than the current state, as it's not entirely apparent what happened when the user taps an `active` `Todo` and completes it.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 5]
```diff
@@ -42,15 +42,18 @@ extension TodosViewController: UITableViewDataSource {
...
+        tableView.performBatchUpdates({
+            var removedTodo = todos[indexPath.section].remove(at: indexPath.row)
+            removedTodo.completed = !removedTodo.completed
+            todos[finishIndexPath.section].insert(removedTodo, at: finishIndexPath.row)
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Lines `47`, `48`, `49` should not reside in the `performBatchUpdates` method - only `UITableView` mutating methods should reside within a `performBatchUpdates` call.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 6]
```diff
@@ -42,15 +42,18 @@ extension TodosViewController: UITableViewDataSource {
...
+            tableView.moveRow(at: indexPath, to: finishIndexPath) },
+                                      completion: { _ in tableView.reloadData() })
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: I would like to see an `animated` reload of the affected `cell`, not a `static` `reloadData()` reload.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 7]
```diff
@@ -53,7 +53,18 @@ extension TodosViewController: UITableViewDelegate {
...
+    }
+    
+    func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
+        return true
+    }
+    
+    func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
+        if editingStyle == UITableViewCellEditingStyle.delete {
+            todos[indexPath.section].remove(at: indexPath.row)
+            tableView.reloadData()
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: I would like to see this `animated`. You can use `UITableView`'s cell-structure mutating methods for this.

**Review by dkulundzic — CHANGES_REQUESTED** [review]
Looks good, but an improvement is necessary.

**Review by dkulundzic — CHANGES_REQUESTED** [review]
Looks good, but an improvement is necessary.

File: Marsiple/App/Todos/TodosViewController.swift [hunk 8]
```diff
@@ -42,13 +42,13 @@ extension TodosViewController: UITableViewDataSource {
...
+        var removedTodo = todos[indexPath.section].remove(at: indexPath.row)
+        removedTodo.completed = !removedTodo.completed
+        todos[finishIndexPath.section].insert(removedTodo, at: finishIndexPath.row)
         tableView.performBatchUpdates({
-            var removedTodo = todos[indexPath.section].remove(at: indexPath.row)
-            removedTodo.completed = !removedTodo.completed
-            todos[finishIndexPath.section].insert(removedTodo, at: finishIndexPath.row)
             tableView.moveRow(at: indexPath, to: finishIndexPath) },
-                                      completion: { _ in tableView.reloadData() })
+                                      completion: { _ in tableView.reloadRows(at: [finishIndexPath], with: .automatic) })
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: Hmm, once again, I have to say that I don't appreciate the code formatting here. 

This is much more readable and concise:
```
tableView.performBatchUpdates({
      tableView.moveRow(at: indexPath, to: finishIndexPath)
}, completion: { _ in
      tableView.reloadRows(at: [finishIndexPath], with: .automatic)
})
```

File: Marsiple/App/Todos/TodosViewController.swift [hunk 9]
```diff
@@ -62,7 +62,7 @@ extension TodosViewController: UITableViewDelegate {
     func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
         if editingStyle == UITableViewCellEditingStyle.delete {
             todos[indexPath.section].remove(at: indexPath.row)
-            tableView.reloadData()
+            tableView.performBatchUpdates({ tableView.deleteRows(at: [indexPath], with: .automatic) }, completion: nil)
         }
```
Comment (dkulundzic) [inline Marsiple/App/Todos/TodosViewController.swift]: As mentioned above, the code formatting could use more readability. 

```
tableView.performBatchUpdates({
      tableView.deleteRows(at: [indexPath], with: .automatic)
}, completion: nil)
```

**Review by dkulundzic — CHANGES_REQUESTED** [review]
The functionality and the UX are there 👍 , but I would like some code formatting changes, to improve readability.

**Review by dkulundzic — APPROVED** [review]
Looks great, lets 🚀

**Comment by matej494** [comment]
Successfully merged branch `todos/view` into branch `todos/main` without conflicts.

**Pull request merged**